enumerable_methods.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    index = 0
+    while index < size
+      yield(self[index])
+      index += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    index = 0
+    while index < size
+      yield(self[index], index)
+      index += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    selected = []
+    my_each do |element|
+      result = yield(element)
+      next unless result
+
+      selected << element
+    end
+    selected
+  end
+
+  def my_all?
+    if block_given?
+      my_each { |element| return false if yield(element) == false }
+    else
+      my_each { |element| return false if element.nil? || element == false }
+    end
+    true
+  end
```

Comment:
There's one more condition missing here. We should be able to pass patterns. See below:

- When a pattern is passed this method should return true if every collection member matches this pattern. Note that pattern could be a `Class` (e.g. `Integer`, `String`) or `RegEx`

   ```ruby
   # When a Class pattern is passed
   irb(main):014:0> [1,2,3].all?(Integer)
   => true
   irb(main):015:0> [1,2,3].my_all?(Integer)
  ArgumentError (wrong number of arguments (given 1, expected 0))

   # When a RegEx pattern is passed
   irb(main):018:0> %w[ant bat cat].all?(/t/)
   => true
   irb(main):019:0> %w[ant bat cat].my_all?(/t/)
   ArgumentError (wrong number of arguments (given 1, expected 0))

   # When a string pattern is passed
   irb(main):018:0> %w[ant bat cat].all?('cat')
   => false
   irb(main):019:0> %w[ant bat cat].my_all?(/t/)
   ArgumentError (wrong number of arguments (given 1, expected 0))
   ```
   Hint: Use ` === ` to check if a member matches a patern.

### Find more details about [`all?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)

#### Remember ([Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421589/))

> As your code will be a bit more complex disable `Style/CaseEquality` cop, learn how to do it here: https://rubydoc.info/gems/rubocop/0.24.1/frames#Disabling_Cops_within_Source_Code